Fix midi rendering of pitch bends

RoadrunnerWMC opened the pull request:
When rendering a pitch bend to MIDI, MuseScore 3 generates a MIDI pitch-wheel event every 16 ticks. However, if the length of the bend isn't a multiple of 16 ticks, MuseScore just stops early, and the bend is left slightly unfinished.

Here's a simple example case:

![image](https://github.com/user-attachments/assets/dc6d0054-ad36-43f2-814c-77a4f26cec2f)

In the current `3.x` branch, the MIDI exported from this score (printed as text with the Python [`mido`](https://mido.readthedocs.io/en/latest/) module) looks like:

<details>
<summary>Mido output</summary>
<pre>
MidiFile(type=1, ticks_per_beat=480, tracks=[
  MidiTrack([
    MetaMessage('time_signature', numerator=4, denominator=4, clocks_per_click=24, notated_32nd_notes_per_beat=8, time=0),
    MetaMessage('set_tempo', tempo=500000, time=0),
    MetaMessage('end_of_track', time=1)]),
  MidiTrack([
    MetaMessage('track_name', name='Piano', time=0),
    MetaMessage('key_signature', key='C', time=0),
    Message('control_change', channel=0, control=121, value=0, time=0),
    Message('control_change', channel=0, control=100, value=0, time=0),
    Message('control_change', channel=0, control=101, value=0, time=0),
    Message('control_change', channel=0, control=6, value=12, time=0),
    Message('control_change', channel=0, control=100, value=127, time=0),
    Message('control_change', channel=0, control=101, value=127, time=0),
    Message('program_change', channel=0, program=0, time=0),
    Message('control_change', channel=0, control=7, value=100, time=0),
    Message('control_change', channel=0, control=10, value=64, time=0),
    Message('control_change', channel=0, control=91, value=0, time=0),
    Message('control_change', channel=0, control=93, value=0, time=0),
    MetaMessage('midi_port', port=0, time=0),
    Message('note_on', channel=0, note=60, velocity=80, time=0),
    Message('pitchwheel', channel=0, pitch=4096, time=0),
    Message('pitchwheel', channel=0, pitch=4096, time=360),
    Message('pitchwheel', channel=0, pitch=3549, time=16),
    Message('pitchwheel', channel=0, pitch=3003, time=16),
    Message('pitchwheel', channel=0, pitch=2457, time=16),
    Message('pitchwheel', channel=0, pitch=1911, time=16),
    Message('pitchwheel', channel=0, pitch=1365, time=16),
    Message('pitchwheel', channel=0, pitch=819, time=16),
    Message('pitchwheel', channel=0, pitch=273, time=16),
    Message('note_on', channel=0, note=60, velocity=0, time=895),
    Message('pitchwheel', channel=0, pitch=0, time=73),
    MetaMessage('end_of_track', time=1)])
])
</pre>
</details>

At the end of the bend, the pitchwheel value is left at 273 until the note ends. (A semitone in this context is 4096/12, so this is about 80 cents sharp.)

This PR adds an extra event at the exact end of the bend if it's necessary. The new MIDI output for this score looks like:

<details>
<summary>Mido output</summary>
<pre>
MidiFile(type=1, ticks_per_beat=480, tracks=[
  MidiTrack([
    MetaMessage('time_signature', numerator=4, denominator=4, clocks_per_click=24, notated_32nd_notes_per_beat=8, time=0),
    MetaMessage('set_tempo', tempo=500000, time=0),
    MetaMessage('end_of_track', time=1)]),
  MidiTrack([
    MetaMessage('track_name', name='Piano', time=0),
    MetaMessage('key_signature', key='C', time=0),
    Message('control_change', channel=0, control=121, value=0, time=0),
    Message('control_change', channel=0, control=100, value=0, time=0),
    Message('control_change', channel=0, control=101, value=0, time=0),
    Message('control_change', channel=0, control=6, value=12, time=0),
    Message('control_change', channel=0, control=100, value=127, time=0),
    Message('control_change', channel=0, control=101, value=127, time=0),
    Message('program_change', channel=0, program=0, time=0),
    Message('control_change', channel=0, control=7, value=100, time=0),
    Message('control_change', channel=0, control=10, value=64, time=0),
    Message('control_change', channel=0, control=91, value=0, time=0),
    Message('control_change', channel=0, control=93, value=0, time=0),
    MetaMessage('midi_port', port=0, time=0),
    Message('note_on', channel=0, note=60, velocity=80, time=0),
    Message('pitchwheel', channel=0, pitch=4096, time=0),
    Message('pitchwheel', channel=0, pitch=4096, time=360),
    Message('pitchwheel', channel=0, pitch=3549, time=16),
    Message('pitchwheel', channel=0, pitch=3003, time=16),
    Message('pitchwheel', channel=0, pitch=2457, time=16),
    Message('pitchwheel', channel=0, pitch=1911, time=16),
    Message('pitchwheel', channel=0, pitch=1365, time=16),
    Message('pitchwheel', channel=0, pitch=819, time=16),
    Message('pitchwheel', channel=0, pitch=273, time=16),
    Message('pitchwheel', channel=0, pitch=0, time=8),
    Message('note_on', channel=0, note=60, velocity=0, time=887),
    Message('pitchwheel', channel=0, pitch=0, time=73),
    MetaMessage('end_of_track', time=1)])
])
</pre>
</details>

Diff:
```diff
@@ -28,7 +28,8 @@
     Message('pitchwheel', channel=0, pitch=1365, time=16),
     Message('pitchwheel', channel=0, pitch=819, time=16),
     Message('pitchwheel', channel=0, pitch=273, time=16),
-    Message('note_on', channel=0, note=60, velocity=0, time=895),
+    Message('pitchwheel', channel=0, pitch=0, time=8),
+    Message('note_on', channel=0, note=60, velocity=0, time=887),
     Message('pitchwheel', channel=0, pitch=0, time=73),
     MetaMessage('end_of_track', time=1)])
 ])
```

The implementation works by increasing the upper bound of the loop to add an additional iteration if the last iteration wouldn't have otherwise landed on `nextPointTick`, and then clamping `i` to `nextPointTick`. I also added a comment to explain what's going on, since it's not very self-explanatory.

Jojo-Schmitz commented:
Rather unsurprisingly a unit test, testBends2, fails and would need to get adjusted.

Jojo-Schmitz commented:
Is that issue also existing in Mu4? If so, do you plan to submit a PR for that too?

RoadrunnerWMC commented:
I'll take a look at the unit test.

Pitch bends seem to be completely different in MuseScore 4. I haven't studied the code, but from experimentation, it adds pitch-wheel events every 10 ticks instead of 16, and if there's a remainder, it adds a final event at the next multiple of 10, even if that's slightly past where the end of the bend should be. For example, a bend the length of a 64th note gets 3 pitch-wheel events (30 total ticks), and a dotted 64th note gets 5 of them (50 total ticks), instead of the 45 total ticks it really should be. That behavior doesn't seem exactly perfect to me, but it's only ever off by 9 midi ticks at most (which is nearly nothing), and it doesn't suffer from the much larger MuseScore 3 issue, so maybe it's fine?

I did accidentally also find that MuseScore 4 generates an apparently invalid midi file if you make a 12-semitone-tall pitch bend, though, so I should probably at least file an issue for that...

Jojo-Schmitz commented:
OK, then once that unit test is fixed, I'll merge this PR

RoadrunnerWMC commented:
The unit test helped me notice that the new implementation would sometimes generate two pitch-wheel events with the same pitch in a row, which isn't ideal. I added some more code to prevent that from happening, which also cleaned up some duplicate events that were already in the reference output for that test.

<details>
<summary>Midi diff</summary>
<pre lang="diff">
@@ -73,7 +73,8 @@
     Message('pitchwheel', channel=0, pitch=901, time=16),
     Message('pitchwheel', channel=0, pitch=1092, time=16),
     Message('pitchwheel', channel=0, pitch=1269, time=16),
-    Message('note_on', channel=0, note=65, velocity=0, time=367),
+    Message('pitchwheel', channel=0, pitch=1365, time=8),
+    Message('note_on', channel=0, note=65, velocity=0, time=359),
     Message('pitchwheel', channel=0, pitch=0, time=1),
     Message('note_on', channel=0, note=65, velocity=80, time=0),
     Message('note_on', channel=0, note=65, velocity=0, time=239),
@@ -223,7 +224,8 @@
     Message('pitchwheel', channel=0, pitch=2116, time=16),
     Message('pitchwheel', channel=0, pitch=2061, time=16),
     Message('note_on', channel=0, note=69, velocity=0, time=3),
-    Message('pitchwheel', channel=0, pitch=0, time=1),
+    Message('pitchwheel', channel=0, pitch=2048, time=1),
+    Message('pitchwheel', channel=0, pitch=0, time=0),
     MetaMessage('end_of_track', time=1)]),
   MidiTrack([
     MetaMessage('track_name', name='Acoustic Guitar', time=0),
@@ -299,8 +301,7 @@
     Message('pitchwheel', channel=1, pitch=614, time=16),
     Message('pitchwheel', channel=1, pitch=641, time=16),
     Message('pitchwheel', channel=1, pitch=682, time=16),
-    Message('pitchwheel', channel=1, pitch=682, time=320),
-    Message('pitchwheel', channel=1, pitch=641, time=16),
+    Message('pitchwheel', channel=1, pitch=641, time=336),
     Message('pitchwheel', channel=1, pitch=614, time=16),
     Message('pitchwheel', channel=1, pitch=573, time=16),
     Message('pitchwheel', channel=1, pitch=546, time=16),
@@ -320,8 +321,7 @@
     Message('pitchwheel', channel=1, pitch=68, time=16),
     Message('pitchwheel', channel=1, pitch=27, time=16),
     Message('pitchwheel', channel=1, pitch=0, time=16),
-    Message('pitchwheel', channel=1, pitch=0, time=320),
-    Message('pitchwheel', channel=1, pitch=27, time=16),
+    Message('pitchwheel', channel=1, pitch=27, time=336),
     Message('pitchwheel', channel=1, pitch=68, time=16),
     Message('pitchwheel', channel=1, pitch=95, time=16),
     Message('pitchwheel', channel=1, pitch=136, time=16),
</pre>
</details>

Jojo-Schmitz commented:
The unit test is still not happy

RoadrunnerWMC commented:
Oops, sorry, I didn't notice that the latest change also made testBends1 stop matching. Should be fixed now.
